File: backend-node/src/routes/rating.js
import express from 'express';
import { pool } from '../db.js';

const router = express.Router();

router.post('/user/select-rating-card', async (req, res, next) => {
  const cardId = String(req.body?.card_id || '');
  if (!/^[0-9a-fA-F-]{36}$/.test(cardId)) return res.status(400).json({ error: 'Valid card_id is required' });
  try {
    const result = await pool.query(
      `UPDATE users u SET active_rating_card_id = c.id
       FROM cards c
       WHERE u.id = $1 AND c.id = $2 AND c.user_id = u.id
       RETURNING u.active_rating_card_id`,
      [req.user.id, cardId]
    );
    if (!result.rows[0]) return res.status(404).json({ error: 'Card not found or does not belong to user' });
    return res.json({ success: true, active_rating_card_id: result.rows[0].active_rating_card_id });
  } catch (error) {
    return next(error);
  }
});

router.get('/rating', async (_req, res, next) => {
  try {
    const result = await pool.query(
      `SELECT u.username, rc.balance
       FROM users u
       LEFT JOIN LATERAL (
         SELECT c.balance
         FROM cards c
         WHERE c.user_id = u.id
         ORDER BY CASE WHEN c.id = u.active_rating_card_id THEN 0 ELSE 1 END, c.created_at ASC
         LIMIT 1
       ) rc ON TRUE
       ORDER BY COALESCE(rc.balance, 0) DESC, u.username ASC
       LIMIT 50`
    );
    return res.json(result.rows.map((row, index) => ({ place: index + 1, username: row.username, balance: row.balance || '0' })));
  } catch (error) {
    return next(error);
  }
});

export default router;

File: backend-python/app/routes/rating.py
import uuid
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import case, func, select
from sqlalchemy.orm import Session
from ..auth import get_current_user
from ..database import get_db
from ..models import Card, User
from ..schemas import RatingBody

router = APIRouter(prefix='/api', tags=['rating'])

@router.post('/user/select-rating-card')
def select_rating_card(body: RatingBody, user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    try:
        card_id = uuid.UUID(body.card_id)
    except ValueError:
        raise HTTPException(400, 'Valid card_id is required')
    with db.begin():
        card = db.scalar(select(Card).where(Card.id == card_id, Card.user_id == user.id).with_for_update())
        if not card:
            raise HTTPException(404, 'Card not found or does not belong to user')
        current = db.scalar(select(User).where(User.id == user.id).with_for_update())
        current.active_rating_card_id = card.id
        db.flush()
    return {'success': True, 'active_rating_card_id': str(card_id)}

@router.get('/rating')
def rating(user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    active_balance = select(Card.balance).where(Card.id == User.active_rating_card_id).scalar_subquery()
    first_balance = select(Card.balance).where(Card.user_id == User.id).order_by(Card.created_at.asc()).limit(1).scalar_subquery()
    balance_expr = func.coalesce(active_balance, first_balance)
    stmt = select(User.username, balance_expr.label('balance')).order_by(balance_expr.desc(), User.username.asc()).limit(50)
    rows = db.execute(stmt).all()
    return [{'place': idx + 1, 'username': row.username, 'balance': str(row.balance or 0)} for idx, row in enumerate(rows)]
